### PR TITLE
support for nested destructions & missing ripple_map server func

### PR DESCRIPTION
**Pull request by leonidaz**
> [!NOTE]
> **Medium Risk**
> Touches Ripple’s compiler analyze/transform pipeline for parameter handling, which can affect generated JS across many components/functions. Changes are well-covered by new client/server tests but could still introduce edge-case regressions in destructuring/type-annotation handling.
> 
> **Overview**
> Adds support for *nested* lazy destructuring (`&{}` / `&[]`) inside otherwise non-lazy component and function parameters by teaching the analyzer to walk subpatterns, propagate relevant TypeScript type annotations (including tuples/rest), and emit lazy transforms for nested patterns.
> 
> Updates both client and server transforms to consistently replace any lazy subpatterns with generated identifiers via a new `replace_lazy_param_pattern()` utility, enabling mixed normal + nested-lazy parameter patterns.
> 
> Also adds missing server runtime helper `ripple_map()` (parity with other collection constructors) and expands test coverage for nested lazy params (including `Tracked<T>` fast-path behavior and rest destructuring), plus minor markdown table formatting cleanup in generated agent docs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ef47e2f8f6c83336ed95d3efbb1d0b9feb406105. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by leonidaz** [comment]
@cursor review

File: packages/ripple/src/compiler/phases/2-analyze/index.js [hunk 1]
```diff
@@ -358,6 +358,50 @@ function is_param_tracked_type(param, context) {
...
+				for (const element of pattern.elements) {
+					if (element !== null) {
+						setup_nested_lazy_param_transforms(element, context);
+					}
+				}
+			}
+
+			return;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (leonidaz) [reply packages/ripple/src/compiler/phases/2-analyze/index.js:598]: @cursor add the fix to this branch

**Comment by leonidaz** [comment]
@cursor add tests for server and client for RestElement for nested lazy destructure that doesn't have the outer lazy destructure.

File: packages/ripple/src/compiler/phases/2-analyze/index.js [hunk 2]
```diff
@@ -358,6 +358,54 @@ function is_param_tracked_type(param, context) {
...
+		case 'RestElement':
+			setup_nested_lazy_param_transforms(pattern.argument, context);
+			return;
+
+		case 'ObjectPattern':
+		case 'ArrayPattern': {
+			if (pattern.lazy) {
+				const param_id = b.id(context.state.scope.generate('lazy'));
+				const is_tracked_type =
+					pattern.type === 'ArrayPattern' && is_param_tracked_type(pattern, context);
```
[the remark this answers is not in the document]
Comment (leonidaz) [reply packages/ripple/src/compiler/phases/2-analyze/index.js]: @cursor ok, add the fix

**Comment by leonidaz** [comment]
@cursor the tests that you added in commit 79dd6d5aafb74a95567ba85a123ff789e7831f88 are not really correct as you're not testing lazy destructure but just a regular one.  you need to be using track and test its result or an object with a getter and setter so you can make sure that destructuring does not remove reactivity (gets direct values) or getters/setters since the lazy destructions go through an intermediary variable.

File: packages/ripple/src/compiler/phases/2-analyze/index.js [hunk 3]
```diff
@@ -330,14 +330,176 @@ function setup_lazy_array_transforms(pattern, source_id, state, writable) {
...
+			if (element_type.type === 'TSRestType') {
+				return element_type.typeAnnotation;
+			}
+		}
+		return undefined;
+	}
+
+	if (index < array_type_annotation.elementTypes.length) {
+		const element_type = normalize_tuple_element_type(array_type_annotation.elementTypes[index]);
+		return element_type.type === 'TSRestType' ? element_type.typeAnnotation : element_type;
```
[the remark this answers is not in the document]
Comment (leonidaz) [reply packages/ripple/src/compiler/phases/2-analyze/index.js]: @cursor fix this in this branch

**Comment by leonidaz** [comment]
@trueadm take a look if you have time.  cursor created types parsing for the fast path and I let it but wondering if it's worth all this, I'll let you decide, didn't realize that's what we were doing.  Initially when I created the pr, I wasn't even thinking about the fast path with function wrappers and was aiming for arr[0], arr[1] notation.

**Comment by trueadm** [comment]
@leonidaz the fast path ensures we don't need to use the 0 access as that has overhead we can avoid in the simple case

**Comment by leonidaz** [comment]
@cursor review

**Review by trueadm — APPROVED** [review]
